TMGToolbox/src/assignment/preprocessing/remove_redundant_functions.py

```diff
@@ -0,0 +1,296 @@
+                        ftlist_net.append(sg_fcn)
+                        
+                for r in trns:
+                    r_fcn = r.penalty_func 
+                    if (r_fcn > 0) and (r_fcn not in fplist_net):
+                        fplist_net.append(r_fcn)
+                        
+            print "There are %s Auto Volume Delay (fd) functions used in all scenarios/networks." %len(fdlist_net)
+            print "There are %s Transit Time (ft) functions used in all scenarios/networks." %len(ftlist_net)
+            print "There are %s Turn Penalty (fp) functions used in all scenarios/networks." %len(fplist_net)
+
+            
+            # compare the functions between networks and the database
+            fdlist_diff = list(set(fdlist_database) - set(fdlist_net))
+            ftlist_diff = list(set(ftlist_database) - set(ftlist_net))
+            fplist_diff = list(set(fplist_database) - set(fplist_net))
+            
+            Redun_fcns = []
+            
+            for fd_i in fdlist_diff:
+                fd_id = "fd%s" %fd_i
+                Redun_fcns.append(_M.emmebank.function(fd_id))
+                
+            for ft_i in ftlist_diff:
+                ft_id = "ft%s" %ft_i
+                Redun_fcns.append(_M.emmebank.function(ft_id))
+                
+            for fp_i in fplist_diff:
+                fp_id = "fp%s" %fp_i
+                Redun_fcns.append(_M.emmebank.function(fp_id))
+
+
+            # export redundant functions
+            if len(Redun_fcns) == 0:
+                print "No redundant function is found."
+            else:
+                _exportFcns(functions = Redun_fcns, export_file = self.ExportFile, append_to_file = False)
+
+            
+            # delete redundant functions from database
+            for redun_i in Redun_fcns:
+                if redun_i is not None:
+                    _deleteFcns(redun_i)
+
+            print "Removed %s functions from the database." %(len(Redun_fcns))
+
+            self.TRACKER.completeTask()
+
+    ##########################################################################################################    
+    
+    #----SUB FUNCTIONS---------------------------------------------------------------------------------  
+    
+    def _GetAtts(self):
+        atts = {
+                "Scenario" : str(self.Scenario.id),
+                "Export File": self.ExportFile,
+                "Version": self.version, 
+                "self": self.__MODELLER_NAMESPACE__}
+            
+        return atts 
+
+    @_m.method(return_type=_m.TupleType)
+    def percent_completed(self):
+        return self.TRACKER.getProgress()
+                
+    @_m.method(return_type=unicode)
+    def tool_run_msg_status(self):
+        return self.tool_run_msg
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
+    
```

Comment:
We can remove all of this extra whitespace.

Reply:
Luna are you familiar with any good formatters? (I am guessing you are using Visual Studio?)